interfaces/seccomp/backend.go

```diff
@@ -174,20 +199,27 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
 	if err := os.MkdirAll(dir, 0755); err != nil {
 		return fmt.Errorf("cannot create directory for seccomp profiles %q: %s", dir, err)
 	}
-	_, _, err = osutil.EnsureDirState(dir, glob, content)
+	changed, removed, err := osutil.EnsureDirState(dir, glob+".src", content)
 	if err != nil {
 		return fmt.Errorf("cannot synchronize security files for snap %q: %s", snapName, err)
 	}
-
-	compiler, err := seccomp_compiler.New(seccompCompilerLookup)
-	if err != nil {
-		return fmt.Errorf("cannot initialize seccomp profile compiler: %v", err)
+	for _, c := range removed {
+		err := os.Remove(bpfBinPath(c))
+		if err != nil && !os.IsNotExist(err) {
+			return err
+		}
 	}
-	for baseName := range content {
-		in := filepath.Join(dirs.SnapSeccompDir, baseName)
-		out := filepath.Join(dirs.SnapSeccompDir, strings.TrimSuffix(baseName, ".src")+".bin")
+	for _, c := range changed {
+		in := bpfSrcPath(c)
+		out := bpfBinPath(c)
+
+		// remove binary profile first
+		err := os.Remove(out)
```

Comment:
I'm somewhat worried about this. Without holding any lock this will open a window when the .bin profile is gone for a split second. Should we perform this whole operation while holding the global lock?

At the very least please include a comment about the special internal wait inside `sc_apply_seccomp_profile_for_security_tag()` in `seccomp-support.c`.

Reply:
thinking further, can't we do the usual thing and generate  the file as a temporary file and rename it to the final name atomically instead of this remove

Reply:
Yeah, we should do that. In case snap-seccomp fails we might consider unlinking the old file (or partially written file in case snap-seccomp is buggy).

When snap-run will see that system key mismatch is present it will spin until that operation is complete and the new system key matches. I would also add this as a comment to the whole block arranging backend operations, to put the reader into the right context.

The consequence of this is that snap-run waits for system-key match, snap-confine waits for the presence of the .bin file to load. Those both have impact on the correctness of the code here.

Reply:
I've talked to @zyga about this after the standup today and the conclusion was that the code is ok without further changes. I have pushed a comment explaining this. To recap:
- when there is a system-key mismatch, `snap run` detects this and waits for snapd, which in turn will only update the system-key when all security backends have finished their setup
- when the compiled profile is missing `snap-confine` polls and waits for `SNAP_CONFINE_MAX_PROFILE_WAIT` time, after which it will fail and die when the profile is still missing
- `snap-seccomp` already uses `osutil.AtomicFileWrite` internally, and the profile is not created when there is a failure compiling the profile

Reply:
Can you add that as a code comment please?

Reply:
There is a commen here: https://github.com/snapcore/snapd/pull/6485/files#diff-a5ed6485262fae2e80ad61a8b843504eR199 and another one just below: https://github.com/snapcore/snapd/pull/6485/files#diff-a5ed6485262fae2e80ad61a8b843504eR228